[iOS] Add watchOS support

xta0 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33318 [iOS] Add watchOS support**

### Summary

Recently, we have a [discussion](https://discuss.pytorch.org/t/libtorch-on-watchos/69073/14) in the forum about watchOS. This PR adds the support for building watchOS  libraries.

### Test Plan

- `BUILD_PYTORCH_MOBILE=1 IOS_PLATFORM=WATCHOS ./scripts/build_ios.sh`

Differential Revision: [D19896534](https://our.internmc.facebook.com/intern/diff/D19896534)

xta0 committed:
[iOS] Add watchOS support

[ghstack-poisoned]

ljk53 commented on CMakeLists.txt:
```diff
@@ -343,7 +343,9 @@ if (INTERN_BUILD_MOBILE AND NOT BUILD_CAFFE2_MOBILE)
   set(NO_API ON)
   set(USE_FBGEMM OFF)
-  set(USE_PYTORCH_QNNPACK ON)
+  if (NOT USE_PYTORCH_QNNPACK)
```
why add this if?

xta0 replied:
because it'll over-write this param in iOS build script. See the change in `build_ios.sh` down below.

ljk53 replied:
I'm not sure if there is a way to differentiate unset variable v.s. OFF variable in cmake. I think even if you explicitly set USE_PYTORCH_QNNPACK to OFF below, "NOT USE_PYTORCH_QNNPACK" will evaluate to TRUE and it will be set to ON again. I haven't tested, though.
I was thinking of not exposing this USE_PYTORCH_QNNPACK flag to external developer to reduce complexity, as it is always ON for mobile builds. But if there is legit reason to disable it for certain mobile devices, maybe we should just remove set(USE_PYTORCH_QNNPACK ON) here.

xta0 replied:
@ljk53 should be `if(NOT DEFINED USE_PYTORCH_QNNPACK)`. My mistake.

xta0 committed:
Update on "[iOS] Add watchOS support"


### Summary

Recently, we have a [discussion](https://discuss.pytorch.org/t/libtorch-on-watchos/69073/14) in the forum about watchOS. This PR adds the support for building watchOS  libraries.

### Test Plan

- `BUILD_PYTORCH_MOBILE=1 IOS_PLATFORM=WATCHOS ./scripts/build_ios.sh`

Differential Revision: [D19896534](https://our.internmc.facebook.com/intern/diff/D19896534)

[ghstack-poisoned]

ljk53 commented on CMakeLists.txt:
```diff
@@ -343,7 +343,9 @@ if (INTERN_BUILD_MOBILE AND NOT BUILD_CAFFE2_MOBILE)
   set(NO_API ON)
   set(USE_FBGEMM OFF)
-  set(USE_PYTORCH_QNNPACK ON)
+  if (NOT DEFINED USE_PYTORCH_QNNPACK)
```
I think we set default value for "USE_PYTORCH_QNNPACK" on Line 181 - will NOT DEFINED ever become true?
I think we should simply removed "set(USE_PYTORCH_QNNPACK ON)" here as we already have default value set to ON for it. I cannot remember why I set this in the first place, maybe was trying to enforce using it everywhere for mobile.

xta0 replied:
That's the best way to do it.

xta0 committed:
Update on "[iOS] Add watchOS support"


### Summary

Recently, we have a [discussion](https://discuss.pytorch.org/t/libtorch-on-watchos/69073/14) in the forum about watchOS. This PR adds the support for building watchOS  libraries.

### Test Plan

- `BUILD_PYTORCH_MOBILE=1 IOS_PLATFORM=WATCHOS ./scripts/build_ios.sh`

Differential Revision: [D19896534](https://our.internmc.facebook.com/intern/diff/D19896534)

[ghstack-poisoned]

ljk53 commented on CMakeLists.txt:
```diff
@@ -343,7 +343,6 @@ if (INTERN_BUILD_MOBILE AND NOT BUILD_CAFFE2_MOBILE)
   set(NO_API ON)
   set(USE_FBGEMM OFF)
-  set(USE_PYTORCH_QNNPACK ON)
   set(USE_QNNPACK OFF)
```
Oh, I think I remember why I added the line above - because I turned USE_QNNPACK OFF and turned USE_PYTORCH_QNNPACK ON - I want to pair them up. i.e. the reason we turned off USE_QNNPACK here is because we turned on better alternative. (just command, no action needed)

ljk53 commented on cmake/Summary.cmake:
```diff
@@ -122,6 +122,7 @@ function (caffe2_print_configuration_summary)
   message(STATUS "  USE_TBB               : ${USE_TBB}")
   message(STATUS "  USE_PROF              : ${USE_PROF}")
   message(STATUS "  USE_QNNPACK           : ${USE_QNNPACK}")
+  message(STATUS "  USE_PYTORCH_QNNPACK   : ${USE_PYTORCH_QNNPACK}")
```
I really hope we can consolidate these two one day :)
